dvc/command/pipeline.py

```diff
@@ -7,178 +7,133 @@
+    return "\n".join(ret)
 
-        for target in self.args.targets:
-            try:
-                if self.args.ascii:
-                    self._show_ascii(
-                        target, self.args.commands, self.args.outs
-                    )
-                elif self.args.dot:
-                    self._write_dot(target, self.args.commands, self.args.outs)
-                elif self.args.tree:
-                    self._show_dependencies_tree(
-                        target, self.args.commands, self.args.outs
-                    )
-                else:
-                    self._show(
-                        target,
-                        self.args.commands,
-                        self.args.outs,
-                        self.args.locked,
-                    )
-            except DvcException:
-                msg = f"failed to show pipeline for '{target}'"
-                logger.exception(msg)
-                return 1
-        return 0
 
+def _build(G, target=None, commands=False, outs=False, full=False):
+    import networkx as nx
+    from dvc.repo.graph import get_pipeline, get_pipelines
+
+    if target:
+        if not full:
+            ancestors = nx.ancestors(G, target)
+            G.remove_nodes_from(ancestors)
+        G = get_pipeline(get_pipelines(G), target)
+
+    def _get(stage):
+        if commands:
+            if not stage.cmd:
+                return []
+            return [stage.cmd]
+        if outs:
+            return stage.outs
+        return [stage.addressing]
+
+    edges = set()
+    for from_stage, to_stage in G.edges():
+        for from_entry in _get(from_stage):
+            for to_entry in _get(to_stage):
+                edges.add((from_entry, to_entry))
+
+    nodes = set()
+    for stage in G.nodes():
+        nodes.update(_get(stage))
 
-class CmdPipelineList(CmdBase):
+    ret = nx.DiGraph()
+    ret.add_edges_from(edges)
+    ret.add_nodes_from(nodes)
+
+    return ret
+
+
+class CmdPipelineShow(CmdBase):
     def run(self):
-        pipelines = self.repo.pipelines
-        for pipeline in pipelines:
-            for stage in pipeline:
-                logger.info(stage.addressing)
-            if len(pipeline) != 0:
-                logger.info("=" * 80)
-        logger.info("{} pipelines total".format(len(pipelines)))
+        try:
+            target = None
+            if self.args.target:
+                stages = self.repo.collect(self.args.target)
+                if len(stages) > 1:
+                    logger.error(
+                        f"'{self.args.target}' contains more than one stage "
+                        "{stages}, please specify one stage"
+                    )
+                    return 1
+                target = stages[0]
+
+            G = _build(
+                self.repo.graph,
+                target=target,
+                commands=self.args.commands,
+                outs=self.args.outs,
+                full=self.args.full,
+            )
```

Comment:
So now we build the DAG once and then visualize it as we please. Previously we used to have different methods to build it, which was too confusing and resulted in different inconsistent behavior for different flags.